Migrate serialization test from boost to catch

jeffin143 opened the pull request:
There you go, We completed all the test and one more boost dependency removed , Refer #2523 

![done_test](https://user-images.githubusercontent.com/21286134/97653960-f8ae1f00-1a87-11eb-876e-59ed6d6e40e7.gif)

birm commented:
Looks like macos is having some issues with the binding tests. Does this look familiar to you?

jeffin143 commented:
> Looks like macos is having some issues with the binding tests. Does this look familiar to you?

There is a discussion ongoing on IRC, any help would be appreciated :)

jeffin143 commented:
Also it is not serialisation test :-> https://github.com/mlpack/mlpack/blob/bf837f950b728bf08b80ccc696c6bc0d296e54af/src/mlpack/bindings/python/CMakeLists.txt#L305

birm reviewed (approved):
I thought it may have been a knock on effect of macos-steps.yaml, but you didn't change that one.
Anyway, this is the end of an era, isn't it :fireworks:

zoq reviewed (commented):
Wow, great job, before we merge this in we should make sure https://github.com/mlpack/mlpack/pull/2678 is merged first, so we don't delete the instructions before that.

jeffin143 commented:
yes @zoq , Also I added some docs on how to run test , and how to run single or all binding related test, let me know if that's not needed, I will remove it

cc @birm  : May be take a second pass over since I added some docs

zoq commented on src/mlpack/tests/README.md:
```diff
@@ -0,0 +1,52 @@
+# Mlpack Tests
```
Let's be consistent and use `mlpack` everywhere.

zoq commented on src/mlpack/tests/README.md:
```diff
@@ -0,0 +1,52 @@
+# Mlpack Tests
+
+This directory contains code and data used to test all the algorthims and function implemented in mlpack.
```
```suggestion
This directory contains code and data used to test all the algorithms and functions implemented in mlpack.
```

zoq commented on src/mlpack/tests/README.md:
```diff
@@ -0,0 +1,52 @@
+
+## Test Directories Structure
+
+- *_test.cpp - All the test realted to algorthims
+- main_tests/*_test.cpp - All binding realted test
+- data - Data needed to run the test
```
```suggestion
- *_test.cpp - methods tests
- main_tests/*_test.cpp - binding tests
- data - data needed to run the tests
```

Simplification?

zoq commented on src/mlpack/tests/README.md:
```diff
@@ -0,0 +1,52 @@
+
+## Building Test
+
+To build test you can simply run `make mlpack_test`.
```
```suggestion
To build the test suite you can simply run `make mlpack_test`.
```

zoq commented on src/mlpack/tests/README.md:
```diff
@@ -0,0 +1,52 @@
+- Adding new functionality.
+- Fixing regressions and bugs.
+
+## Building Test
```
```suggestion
## Building Tests
```

zoq commented on src/mlpack/tests/README.md:
```diff
@@ -0,0 +1,52 @@
+
+To build test you can simply run `make mlpack_test`.
+
+## To run Test 
```
```suggestion
## To run Tests
```

zoq commented on src/mlpack/tests/README.md:
```diff
@@ -0,0 +1,52 @@
+
+## To run Test 
+
+We use `Catch2` to write our test and to run the test, you can simply run
```
```suggestion
We use `Catch2` to write our tests, to run all tests, you can simply run:
```

zoq commented on src/mlpack/tests/README.md:
```diff
@@ -0,0 +1,52 @@
+
+We use `Catch2` to write our test and to run the test, you can simply run
+
+` ./bin/mlpack_test`
```
```suggestion
`./bin/mlpack_test`
```

Remove extra space.

zoq commented on src/mlpack/tests/README.md:
```diff
@@ -0,0 +1,52 @@
+
+` ./bin/mlpack_test`
+
+Also to run all the test in a particular file you can issue a command as
```
```suggestion
Also to run all tests in a particular file you can run:
```

zoq commented on src/mlpack/tests/README.md:
```diff
@@ -0,0 +1,52 @@
+
+`./bin/mlpack_test "[testname]"`
+
+for example to run all the test which is in `cv_test.cpp` file you can issue a command as
```
```suggestion
where `testname` is the name of the test suite. For example to run all
collaborative filtering tests implemented in `cv_test.cpp` you can run:
```

Just a suggestion.

zoq commented on src/mlpack/tests/README.md:
```diff
@@ -0,0 +1,52 @@
+
+`./bin/mlpack_test "[CVTest]"`
+
+Now similary you can run all the binding related test as 
```
```suggestion
Now similarly you can run all the binding related tests using:
```

Minor spelling fix.

zoq commented on src/mlpack/tests/README.md:
```diff
@@ -0,0 +1,52 @@
+Now similary you can run all the binding related test as 
+
+`./bin/mlpack_test "[BindingTests]"`
+
+To run a single test you can explicitly give the name of the test, for example to run
+`BinaryClassificationMetricsTest` present in `cv_test.cpp` file you can run the following
+command:
```
```suggestion

To run a single test, you can explicitly provide the name of the test, for example, to run
`BinaryClassificationMetricsTest` implemented in `cv_test.cpp` you can run the following:
```

zoq commented on src/mlpack/tests/README.md:
```diff
@@ -0,0 +1,52 @@
+
+`./bin/mlpack_test BinaryClassificationMetricsTest`
+
+Catch2 also allows very other command which you can view using:
```
```suggestion
Catch2 provides many other features like filter, checkout the [Catch2 reference section](docs/Readme.md#top) - for more details.
```

With that, I would delete the section below.

zoq reviewed (commented):
Just some suggestions for the documentation.

birm commented on src/mlpack/tests/README.md:
```diff
@@ -0,0 +1,46 @@
+
+## To run Tests
+
+We use `Catch2` to write our tests, to run all tests, you can simply run:
```
```suggestion
We use `Catch2` to write our tests. To run all tests, you can simply run:
```

birm commented on src/mlpack/tests/README.md:
```diff
@@ -0,0 +1,46 @@
+
+`./bin/mlpack_test`
+
+Also to run all tests in a particular file you can run:
```
```suggestion
To run all tests in a particular file you can run:
```

birm commented on src/mlpack/tests/README.md:
```diff
@@ -0,0 +1,46 @@
+
+## Add tests 
+
+We have a rich test suite, consisting of almost 2000 test and still counting. So it is suggested to add tests when:
```
```suggestion
We have a rich test suite, consisting of almost 2000 tests (and still counting). It is suggested to add tests when:
```

birm reviewed (approved):
Minor doc readability suggestions, but the core still seems good :)

Aakash-kaushik commented:
Goodbye to Boost unit_test_framework.

zoq reviewed (approved):
Wow, great job everyone.

![tenor](https://user-images.githubusercontent.com/4209744/98977976-87577d00-2519-11eb-87d5-25059fd1b481.gif)
